Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp

```diff
@@ -110,36 +211,488 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+
+                    auto emptyCallFrames = JSON::ArrayOf<BidiScript::StackFrame>::create();
+                    auto stackTrace = BidiScript::StackTrace::create()
+                        .setCallFrames(WTFMove(emptyCallFrames))
+                        .release();
+
+                    exceptionDetails = BidiScript::ExceptionDetails::create()
+                        .setLineNumber(0)
+                        .setColumnNumber(0)
+                        .setText("Malformed envelope result"_s)
+                        .setException(WTFMove(exceptionRemote))
+                        .setStackTrace(WTFMove(stackTrace))
+                        .release();
+                }
+            }
+        } else {
+            evalType = BidiScript::EvaluateResultType::Exception;
+
+            auto exceptionRemote = BidiScript::RemoteValue::create()
+                .setType(BidiScript::RemoteValueType::Object)
+                .release();
+            exceptionRemote->setValue(JSON::Value::create(String("Malformed envelope"_s)));
+
+            auto emptyCallFrames = JSON::ArrayOf<BidiScript::StackFrame>::create();
+            auto stackTrace = BidiScript::StackTrace::create()
+                .setCallFrames(WTFMove(emptyCallFrames))
+                .release();
+
+            exceptionDetails = BidiScript::ExceptionDetails::create()
+                .setLineNumber(0)
+                .setColumnNumber(0)
+                .setText("Malformed envelope result"_s)
+                .setException(WTFMove(exceptionRemote))
+                .setStackTrace(WTFMove(stackTrace))
+                .release();
+        }
+    }
+
+    if (evalType == BidiScript::EvaluateResultType::Exception) {
+        if (!exceptionDetails) {
+            auto fallbackRemote = BidiScript::RemoteValue::create()
+                .setType(BidiScript::RemoteValueType::Error)
+                .release();
+            fallbackRemote->setValue(JSON::Value::create(String("Script evaluation failed"_s)));
+
+            auto emptyCallFrames = JSON::ArrayOf<BidiScript::StackFrame>::create();
+            auto stackTrace = BidiScript::StackTrace::create()
+                .setCallFrames(WTFMove(emptyCallFrames))
+                .release();
+
+            exceptionDetails = BidiScript::ExceptionDetails::create()
+                .setLineNumber(0)
+                .setColumnNumber(0)
+                .setText("JavaScript exception occurred"_s)
+                .setException(WTFMove(fallbackRemote))
+                .setStackTrace(WTFMove(stackTrace))
+                .release();
+        }
+        callback({ { evalType, realmId, nullptr, WTFMove(exceptionDetails) } });
+        return;
+    }
+
+    callback({ { evalType, realmId, WTFMove(remote), nullptr } });
 }
 
+RefPtr<Inspector::Protocol::BidiScript::RemoteValue> BidiScriptAgent::serializeAsRemoteValue(RefPtr<JSON::Value> value)
+{
```

Comment:
It would be good to add a link to the spec: https://w3c.github.io/webdriver-bidi/#serialize-as-a-remote-value